File: simulator/coordinator/tests/testdata/multi_language/m1.py
import json

from modeling import (
    Atomic,
    RunnableModelCfg,
    RunnableModelPortCfg,
    new_atomic_from_cfg,
)


class PythonAdder(Atomic):
    def __init__(self, id: str, name: str, ports=None):
        super().__init__(id=id, name=name, ports=ports)
        self.value = 0
        self.add_value = 10
        self.result = 0
        self.has_result = False

    def initialize(self) -> None:
        self.has_result = False
        self.passivate()

    def exit(self) -> None:
        pass

    def delt_int(self) -> None:
        self.passivate()

    def delt_ext(self, e: float) -> None:
        try:
            in_port = self.get_port_by_name("in")
            values = in_port.get_values()
            if values:
                data = json.loads(values[0])
                input_value = data.get("value", 0)
                self.result = input_value + self.add_value
                self.has_result = True
        except (KeyError, json.JSONDecodeError, IndexError):
            pass
        self.passivate()

    def delt_con(self, e: float) -> None:
        self.delt_ext(e)

    def lambda_(self) -> None:
        if not self.has_result:
            return

        try:
            out_port = self.get_port_by_name("out")
        except KeyError:
            return

        payload = {"value": self.result}
        out_port.add_value(payload)
        self.has_result = False


def NewModel(config: dict) -> Atomic:
    raw_ports = config.get("ports") or []
    ports_cfg = [
        RunnableModelPortCfg(id=p["id"], type=p["type"])
        for p in raw_ports
    ]

    cfg = RunnableModelCfg(
        id=config["id"],
        name=config["name"],
        ports=ports_cfg,
    )

    return new_atomic_from_cfg(cfg, PythonAdder)

File: simulator/coordinator/internal/simulation/testdata/multi_language/m1.go
package main

import (
	modeling "devsforge-wrapper/modeling"
	"encoding/json"
)

type GoSender struct {
	modeling.Atomic
	InitialValue int
	Sent         bool
}

func NewModel(cfg modeling.RunnableModel) modeling.Atomic {
	return &GoSender{
		Atomic:       modeling.NewAtomic(cfg),
		InitialValue: 5,
		Sent:         false,
	}
}

func (m *GoSender) Initialize() {
	m.Sent = false
	m.HoldIn("active", 0.0)
}

func (m *GoSender) Exit() {}

func (m *GoSender) DeltInt() {
	m.Passivate()
}

func (m *GoSender) DeltExt(e float64) {
	m.Continue(e)
}

func (m *GoSender) DeltCon(e float64) {
	m.DeltInt()
}

func (m *GoSender) Lambda() {
	if m.Sent {
		m.Passivate()
		return
	}

	outPort, err := m.GetPortByName("out")
	if err != nil {
		return
	}

	payload, err := json.Marshal(map[string]int{
		"value": m.InitialValue,
	})
	if err != nil {
		return
	}

	outPort.AddValue(payload)
	m.Sent = true
	m.HoldIn("active", 1.0)
}
